mhkit/loads/extreme.py

```diff
@@ -0,0 +1,361 @@
+    # Divide into seven sets & fit Weibull
+    subset_shape_params = np.zeros(7)
+    subset_scale_params = np.zeros(7)
+    setLim = np.arange(0.60, 0.90, 0.05)
+    func = lambda x, c, s: stats.exponweib(a=1, c=c, loc=0, scale=s).cdf(x)
+    for set in range(7):
+        xset = x[(F > setLim[set])]
+        Fset = F[(F > setLim[set])]
+        popt, _ = optimize.curve_fit(func, xset, Fset, p0=p0)
+        subset_shape_params[set] = popt[0]
+        subset_scale_params[set] = popt[1]
+    # peaks distribution
+    peaks_params = [1, np.mean(subset_shape_params), 0,
+                    np.mean(subset_scale_params)]
+    param_names = ['a', 'c', 'loc', 'scale']
+    peaks_params = {k: v for k, v in zip(param_names, peaks_params)}
+    peaks = stats.exponweib(**peaks_params)
+    # save the parameter info
+    peaks.params = peaks_params
+    peaks.subset_shape_params = subset_shape_params
+    peaks.subset_scale_params = subset_scale_params
+    return peaks
+
+
+def peaks_distribution_peaksOverThreshold(x, threshold):
+    """ Estimate the peaks distribution using the peaks over threshold
+    method.
+
+    This fits a generalized Pareto distribution to all the peaks above
+    the specified threshold.
+    The distribution is only defined for values above the threshold
+    and therefore cannot be used to obtain integral metrics such as the
+    expected value.
+    A typical choice of threshold is 1.4 standard deviations above the
+    mean.
+    The peaks over threshold distribution can be accessed through the
+    `pot` field of the returned peaks distribution.
+
+    Parameters
+    ----------
+        x : np.array
+            Global peaks.
+
+    Returns
+    -------
+        peaks: scipy.stats rv_frozen
+            Probability distribution of the peaks.
+    """
+    # peaks over threshold
+    x = np.sort(x)
+    pot = x[(x > threshold)] - threshold
+    N = len(x)
+    Npot = len(pot)
+    # Fit a generalized Pareto
+    pot_params = stats.genpareto.fit(pot, floc=0.)
+    param_names = ['c', 'loc', 'scale']
+    pot_params = {k: v for k, v in zip(param_names, pot_params)}
+    pot = stats.genpareto(**pot_params)
+    # save the parameter info
+    pot.params = pot_params
+    # peaks
+    class _Peaks(stats.rv_continuous):
+
+        def __init__(self, *args, **kwargs):
+            self.pot = kwargs.pop('pot_distribution')
+            self.threshold = kwargs.pop('threshold')
+            super().__init__(*args, **kwargs)
+
+        def _cdf(self, x):
+            x = np.atleast_1d(np.array(x))
+            out = np.zeros(x.shape)
+            out[x < self.threshold] = np.NaN
+            xt = x[x >= self.threshold]
+            if xt.size != 0:
+                pot_ccdf = 1. - self.pot.cdf(xt-self.threshold)
+                prop_pot = Npot/N
+                out[x >= self.threshold] = 1. - (prop_pot * pot_ccdf)
+            return out
+
+    peaks = _Peaks(name="peaks", pot_distribution=pot, threshold=threshold)
```

Comment:
Interesting use of a class inside a function. Is there a way to simplify this or is this the best way to do the computations?

Reply:
For some methods you simply fit a specific distribution (eg. Weibull). So you can  return a `scipy.stats distribution`. This is very convenient, you can do a lot with it (probability functions: PDF, CDF, PPF, etc; and metrics: expected value, etc). Other methods, however, are not as straightforward. The easiest solution I came up with was to subclass the `scipy.stats.rv_continuous` which can be  done by defining one of the PDF or CDF only, and you then get access to all the functionalities  of `scipy.stats distributions`. Else we would need to write our own implementation of these different functions and metrics. We should discuss this  more. 

- [Info on subclassing the `scipy.stats.rv_continuous`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html)

Reply:
@ssolson we discussed this and think that subclassing `scipy.stats.rv_continuous` ([see documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html)) is acceptable here. What do  you think? Let me know if you want to discuss.